### PR TITLE
Fail Driver on exception from processNewSources

### PR DESCRIPTION
**Pull request by losipiuk**
Exceptions thrown from processNewSources were not handled correctly. We just logged an error and marked the PrioritizedSplitRunner as finished. The for which the PSR were executing remained RUNNING and query hang indefinitely.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# General
* Fix rare case when queries could hang.
```

### CODE REVIEW

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
CI
- https://github.com/trinodb/trino/issues/20326
- and something else ([pt (default, suite-iceberg, )](https://github.com/trinodb/trino/actions/runs/8554759250/job/23441149172?pr=21397#logs))

will retry

File: core/trino-main/src/main/java/io/trino/operator/Driver.java [hunk 1]
```diff
@@ -220,43 +220,51 @@ public void updateSplitAssignment(SplitAssignment splitAssignment)
     @GuardedBy("exclusiveLock")
     private void processNewSources()
     {
-        checkLockHeld("Lock must be held to call processNewSources");
+        try {
+            checkLockHeld("Lock must be held to call processNewSources");
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/operator/Driver.java:224]: Move this outside the try

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/operator/Driver.java [hunk 2]
```diff
@@ -220,43 +220,51 @@ public void updateSplitAssignment(SplitAssignment splitAssignment)
...
 
-        currentSplitAssignment = newAssignment;
+            currentSplitAssignment = newAssignment;
+        }
+        catch (Throwable failure) {
+            driverContext.failed(failure);
+            throwIfUnchecked(failure);
+            // should never happen
+            throw new AssertionError(failure);
+        }
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:267]: The meaning of that was just we should never reach throw new `AssertionError` as `failure` should be unchecked. I copied this pattern from `tryWithLock` - but it is not necessary. dropping.

**Comment by losipiuk** [comment]
Thanks for review. 
The change to be effective required also changes in `TimeSharingTaskExecutor`. Updated

File: core/trino-main/src/main/java/io/trino/execution/executor/timesharing/TimeSharingTaskExecutor.java [hunk 3]
```diff
@@ -593,10 +593,10 @@ public void run()
                     catch (Throwable t) {
                         if (!split.isDestroyed()) { // ignore random errors due to driver thread interruption
                             if (t instanceof TrinoException trinoException) {
-                                log.error(t, "Error processing %s: %s: %s", split.getInfo(), trinoException.getErrorCode().getName(), trinoException.getMessage());
+                                log.debug(t, "Error processing %s: %s: %s", split.getInfo(), trinoException.getErrorCode().getName(), trinoException.getMessage());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/executor/timesharing/TimeSharingTaskExecutor.java:596]: in `Decrease logging verbocity`

verbocity -> verbosity

maybe add more color to commit title (like "Decrease logging verbosity for split processing failures")

noun missing in "Getting exception in ??? is a non-rare faililure mode."

> Also excpetion will not be lost as it is propagated to split runner and
eventually will surfase as task failure and also possibly query failure.

Isn't that best effort?

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/Driver.java [hunk 4]
```diff
@@ -221,42 +221,47 @@ public void updateSplitAssignment(SplitAssignment splitAssignment)
     private void processNewSources()
     {
         checkLockHeld("Lock must be held to call processNewSources");
+        try {
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/operator/Driver.java:224]: The body of commit message was here by incident. Copied from other commit.

File: core/trino-main/src/main/java/io/trino/execution/executor/timesharing/TimeSharingTaskExecutor.java [hunk 5]
```diff
@@ -600,6 +600,7 @@ public void run()
                                 log.error(t, "Error processing %s", split.getInfo());
                             }
                         }
+                        split.markFailed(t);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/executor/timesharing/TimeSharingTaskExecutor.java:603]: Fixed commit message - not sure about the comment - as I am not sure it really is best effort.

**Comment by losipiuk** [comment]
CI failure timeouts in `TestMergeOperator`. Cannot reproduce (with high concurrency and repeat count). Does not seem related.

**Comment by losipiuk** [comment]
@martint PTAL

**Comment by losipiuk** [comment]
Addressed comments. @martint let me know if I should improve something in followup.+

**Pull request merged**